ctags/main/entry.c

```diff
@@ -377,38 +508,948 @@ static int replacementTruncate (const char *const name, const long size)
+	if (slot->extensionFields.varType)
+		eFree ((char *)slot->extensionFields.varType);
+/* GEANY DIFF END */
+#ifdef HAVE_LIBXML
+	if (slot->extensionFields.xpath)
+		eFree ((char *)slot->extensionFields.xpath);
+#endif
+
+	if (slot->sourceLanguage)
+		eFree ((char *)slot->sourceLanguage);
+	if (slot->sourceFileName)
+		eFree ((char *)slot->sourceFileName);
+
+	clearParserFields (slot);
+}
+
+static unsigned int queueTagEntry(const tagEntryInfo *const tag)
+{
+	unsigned int i;
+	void *tmp;
+	tagEntryInfo * slot;
+
+	if (! (TagFile.corkQueue.count < TagFile.corkQueue.length))
+	{
+		if (!TagFile.corkQueue.length)
+			TagFile.corkQueue.length = 1;
+
+		tmp = eRealloc (TagFile.corkQueue.queue,
+				sizeof (*TagFile.corkQueue.queue) * TagFile.corkQueue.length * 2);
+
+		TagFile.corkQueue.length *= 2;
+		TagFile.corkQueue.queue = tmp;
+	}
+
+	i = TagFile.corkQueue.count;
+	TagFile.corkQueue.count++;
+
+
+	slot = TagFile.corkQueue.queue + i;
+	recordTagEntryInQueue (tag, slot);
+
+	return i;
+}
+
+
+static void *writerData;
+static tagWriter *writer;
+
+extern void setTagWriter (tagWriter *t)
+{
+	writer = t;
+}
+
+extern bool outpuFormatUsedStdoutByDefault (void)
+{
+	return writer->useStdoutByDefault;
+}
+
+extern void setupWriter (void)
+{
+	if (writer->preWriteEntry)
+		writerData = writer->preWriteEntry (TagFile.mio);
+	else
+		writerData = NULL;
+}
+
+extern void teardownWriter (const char *filename)
+{
+	if (writer->postWriteEntry)
+		writer->postWriteEntry (TagFile.mio, filename, writerData);
+}
+
+static void buildFqTagCache (const tagEntryInfo *const tag)
+{
+	renderFieldEscaped (FIELD_SCOPE_KIND_LONG, tag, NO_PARSER_FIELD);
+	renderFieldEscaped (FIELD_SCOPE, tag, NO_PARSER_FIELD);
+}
+
+#ifdef CTAGS_LIB
+static void initCtagsTag(ctagsTag *tag, const tagEntryInfo *info)
+{
+	tag->name = info->name;
+	tag->signature = info->extensionFields.signature;
+	tag->scopeName = info->extensionFields.scopeName;
+	tag->inheritance = info->extensionFields.inheritance;
+	tag->varType = info->extensionFields.varType;
+	tag->access = info->extensionFields.access;
+	tag->implementation = info->extensionFields.implementation;
+	tag->kindLetter = info->kind->letter;
+	tag->isFileScope = info->isFileScope;
+	tag->lineNumber = info->lineNumber;
+}
+#endif
+
+static void writeTagEntry (const tagEntryInfo *const tag)
+{
+	int length = 0;
+
+	if (tag->placeholder)
+		return;
+	if (! tag->kind->enabled)
+		return;
+	if (tag->extensionFields.roleIndex != ROLE_INDEX_DEFINITION
+	    && ! isXtagEnabled (XTAG_REFERENCE_TAGS))
+		return;
+
+	DebugStatement ( debugEntry (tag); )
+	Assert (writer);
+
+	if (includeExtensionFlags ()
+	    && isXtagEnabled (XTAG_QUALIFIED_TAGS)
+	    && doesInputLanguageRequestAutomaticFQTag ())
+		buildFqTagCache (tag);
+
+#ifdef CTAGS_LIB
+	getTagScopeInformation((tagEntryInfo *)tag, NULL, NULL);
+
+	if (TagEntryFunction != NULL)
+	{
+		ctagsTag t;
+
+		initCtagsTag(&t, tag);
+		length = TagEntryFunction(&t, TagEntryUserData);
+	}
+#else
```

Comment:
Couldn't we simply have our own writer?  Might be more work, but maybe less divergence?